dev/com.ibm.ws.microprofile.config.1.2/src/com/ibm/ws/microprofile/config12/converters/ImplicitConverter.java

```diff
@@ -28,117 +33,196 @@
+        return getFunction(reflectionClass, "parse", CharSequence.class);
     }
 
     @FFDCIgnore(NoSuchMethodException.class)
     @Trivial
-    private static Method getParse(Class<?> reflectionClass) {
-        Method method = null;
+    protected static Function<String, ?> getFunction(Class<?> reflectionClass, String methodName, Class<?>... paramTypes) {
+        Function<String, ?> implicitFunction = null;
         try {
-            method = reflectionClass.getMethod("parse", CharSequence.class);
+            Method method = reflectionClass.getMethod(methodName, paramTypes);
             if ((method.getModifiers() & Modifier.STATIC) == 0) {
                 method = null;
             } else if (!reflectionClass.equals(method.getReturnType())) {
                 method = null;
             }
+
+            if (method != null) {
+                implicitFunction = wrapMethod(reflectionClass, method);
+            }
         } catch (NoSuchMethodException e) {
             //No FFDC
         }
-        return method;
+        return implicitFunction;
+    }
+
+    @Trivial
+    protected static <X> Function<String, X> wrapMethod(Class<X> reflectionClass, Method method) {
+        return new MethodFunction<X>(reflectionClass, method);
+    }
+
+    @Trivial
+    protected static <X> Function<String, X> wrapConstructor(Class<X> reflectionClass, Constructor<X> ctor) {
+        return new ConstructorFunction<X>(reflectionClass, ctor);
     }
 
     /** {@inheritDoc} */
     @Override
     public Object convert(String value) {
-        Object converted = null;
-        if (value != null) { //if the value is null then we always return null
-            try {
-                if (this.ctor != null) {
-                    converted = this.ctor.newInstance(value);
-                } else if (this.valueOfMethod != null) {
-                    converted = this.valueOfMethod.invoke(null, value);
-                } else if (this.parseMethod != null) {
-                    converted = this.parseMethod.invoke(null, value);
-                }
-            } catch (InvocationTargetException e) {
-                Throwable cause = e.getCause();
-                if (cause instanceof IllegalArgumentException) {
-                    throw (IllegalArgumentException) cause;
-                } else {
-                    throw new ConversionException(cause);
-                }
-            } catch (IllegalAccessException | InstantiationException e) {
-                throw new ConversionException(e);
-            }
-        }
-        return converted;
+        return this.implicitFunction.apply(value);
     }
 
     /** {@inheritDoc} */
     @Override
     public String toString() {
         return "Implicit Converter for type " + getType();
     }
+
+    protected static class ConstructorFunction<X> implements Function<String, X> {
+        private final static Map<Class<?>, Constructor<?>> constructorCache = Collections.synchronizedMap(new WeakHashMap<>());
```

Comment:
I guess this comment is more important in the Manager if we don't keep the constructor / method ones.

Reply:
I don't think there is such a thing as a ConcurrentWeakHashMap??

You told me that if the ImplicitConverter had a reference to the Constructor or Methods then the WeakHashMap cache in the Manager would not be enough. The Method would hold a reference to the Class and the WeakReference in the cache would never get removed?

Reply:
There is not a ConcurrentWeakHashMap.  You can easily use a ConcurrentHashMap with a WeakReference impl key.  I showed examples of how to do that via slack earlier.  

Right now I see you look up ever time and didn't have a reference to the Method / Constructor in the function which is what I thought it was doing because that would make sense.  Instead I would suggest that the Function have a WeakReference to the Method / Constructor and you just do a get on the WeakReference to get the Method / Constructor instead of using yet another WeakHashMap which wouldn't work anyway because the value is the Constructor / Method which will keep a hard reference on the Class.  The only way to use the WeakHashMap would be to create a WeakReference on the Constructor / Method to use it as the value so it is best just to not use the WeakHashMap.

Reply:
The WeakHashMap stuff was too hard to make work efficiently. So I gave up on the static caches and went with a simple cache on the `Config12ConversionManager`. This is scoped by the Config itself so when the application is stopped, the Config is removed and this cache will go too.